### ISSUE TITLE
Void Linux Build Error

### ISSUE THREAD

**Issue by 13-05** [rustformers/llama-rs#82]
### System Info
**OS** -> Void Linux (x86_64) (glibc), linux kernel 6.1.21_1
**rustc** -> rustc 1.68.1 (8460ca823 2023-03-20)
**cargo** -> cargo 1.68.1 (115f34552 2023-02-26)

### Are the required C compilers present?
I have clang version 12.0.1, and gcc version 12.2.0 installed.

### What error does the installation give?
[Here's the entire log, uploaded to termbin.com for convenience.](https://termbin.com/mb48)
This looks like the important bit:
```
The following warnings were emitted during compilation:

warning: In file included from /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/immintrin.h:107,
warning:                  from ggml/ggml.c:155:
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h: In function 'ggml_vec_dot_f16':
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h:52:1: error: inlining failed in call to 'always_inline' '_mm256_cvtph_ps': target specific option mismatch
warning:    52 | _mm256_cvtph_ps (__m128i __A)
warning:       | ^~~~~~~~~~~~~~~
warning: ggml/ggml.c:916:33: note: called from here
warning:   916 | #define GGML_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
warning:       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ggml/ggml.c:926:37: note: in expansion of macro 'GGML_F32Cx8_LOAD'
warning:   926 | #define GGML_F16_VEC_LOAD(p, i)     GGML_F32Cx8_LOAD(p)
warning:       |                                     ^~~~~~~~~~~~~~~~
warning: ggml/ggml.c:1279:21: note: in expansion of macro 'GGML_F16_VEC_LOAD'
warning:  1279 |             ay[j] = GGML_F16_VEC_LOAD(y + i + j*GGML_F16_EPR, j);
warning:       |                     ^~~~~~~~~~~~~~~~~
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h:52:1: error: inlining failed in call to 'always_inline' '_mm256_cvtph_ps': target specific option mismatch
warning:    52 | _mm256_cvtph_ps (__m128i __A)
warning:       | ^~~~~~~~~~~~~~~
warning: ggml/ggml.c:916:33: note: called from here
warning:   916 | #define GGML_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
warning:       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ggml/ggml.c:926:37: note: in expansion of macro 'GGML_F32Cx8_LOAD'
warning:   926 | #define GGML_F16_VEC_LOAD(p, i)     GGML_F32Cx8_LOAD(p)
warning:       |                                     ^~~~~~~~~~~~~~~~
warning: ggml/ggml.c:1278:21: note: in expansion of macro 'GGML_F16_VEC_LOAD'
warning:  1278 |             ax[j] = GGML_F16_VEC_LOAD(x + i + j*GGML_F16_EPR, j);
warning:       |                     ^~~~~~~~~~~~~~~~~
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h:52:1: error: inlining failed in call to 'always_inline' '_mm256_cvtph_ps': target specific option mismatch
warning:    52 | _mm256_cvtph_ps (__m128i __A)
warning:       | ^~~~~~~~~~~~~~~
warning: ggml/ggml.c:916:33: note: called from here
warning:   916 | #define GGML_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
warning:       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ggml/ggml.c:926:37: note: in expansion of macro 'GGML_F32Cx8_LOAD'
warning:   926 | #define GGML_F16_VEC_LOAD(p, i)     GGML_F32Cx8_LOAD(p)
warning:       |                                     ^~~~~~~~~~~~~~~~
warning: ggml/ggml.c:1278:21: note: in expansion of macro 'GGML_F16_VEC_LOAD'
warning:  1278 |             ax[j] = GGML_F16_VEC_LOAD(x + i + j*GGML_F16_EPR, j);
warning:       |                     ^~~~~~~~~~~~~~~~~
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h:52:1: error: inlining failed in call to 'always_inline' '_mm256_cvtph_ps': target specific option mismatch
warning:    52 | _mm256_cvtph_ps (__m128i __A)
warning:       | ^~~~~~~~~~~~~~~
warning: ggml/ggml.c:916:33: note: called from here
warning:   916 | #define GGML_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
warning:       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ggml/ggml.c:926:37: note: in expansion of macro 'GGML_F32Cx8_LOAD'
warning:   926 | #define GGML_F16_VEC_LOAD(p, i)     GGML_F32Cx8_LOAD(p)
warning:       |                                     ^~~~~~~~~~~~~~~~
warning: ggml/ggml.c:1279:21: note: in expansion of macro 'GGML_F16_VEC_LOAD'
warning:  1279 |             ay[j] = GGML_F16_VEC_LOAD(y + i + j*GGML_F16_EPR, j);
warning:       |                     ^~~~~~~~~~~~~~~~~
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h:52:1: error: inlining failed in call to 'always_inline' '_mm256_cvtph_ps': target specific option mismatch
warning:    52 | _mm256_cvtph_ps (__m128i __A)
warning:       | ^~~~~~~~~~~~~~~
warning: ggml/ggml.c:916:33: note: called from here
warning:   916 | #define GGML_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
warning:       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ggml/ggml.c:926:37: note: in expansion of macro 'GGML_F32Cx8_LOAD'
warning:   926 | #define GGML_F16_VEC_LOAD(p, i)     GGML_F32Cx8_LOAD(p)
warning:       |                                     ^~~~~~~~~~~~~~~~
warning: ggml/ggml.c:1278:21: note: in expansion of macro 'GGML_F16_VEC_LOAD'
warning:  1278 |             ax[j] = GGML_F16_VEC_LOAD(x + i + j*GGML_F16_EPR, j);
warning:       |                     ^~~~~~~~~~~~~~~~~
warning: /usr/lib/gcc/x86_64-unknown-linux-gnu/12.2/include/f16cintrin.h:52:1: error: inlining failed in call to 'always_inline' '_mm256_cvtph_ps': target specific option mismatch
warning:    52 | _mm256_cvtph_ps (__m128i __A)
warning:       | ^~~~~~~~~~~~~~~
warning: ggml/ggml.c:916:33: note: called from here
warning:   916 | #define GGML_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
warning:       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ggml/ggml.c:926:37: note: in expansion of macro 'GGML_F32Cx8_LOAD'
warning:   926 | #define GGML_F16_VEC_LOAD(p, i)     GGML_F32Cx8_LOAD(p)
warning:       |                                     ^~~~~~~~~~~~~~~~
warning: ggml/ggml.c:1279:21: note: in expansion of macro 'GGML_F16_VEC_LOAD'
warning:  1279 |             ay[j] = GGML_F16_VEC_LOAD(y + i + j*GGML_F16_EPR, j);
warning:       |                     ^~~~~~~~~~~~~~~~~

error: failed to run custom build command for `ggml-raw v0.1.0 (/home/outsider/.cargo/git/checkouts/llama-rs-962022d29f37c95e/a067431/ggml-raw)`
```

Should this be reported to the [base repository, ggml](https://github.com/ggerganov/ggml), as it's a compilation error for that project, or is it still relevant here?

**Comment by philpax** [comment]
Hm, this could be an issue on our end depending on build settings. Can you check if `llama.cpp` builds for you with its Makefile? If it does, then it's something we'll need to track down in our build script.

**Comment by 13-05** [comment]
Thanks for the swift response! Unfortunately, `llama.cpp` doesn't build for me; it gives the same errors. Is this an issue on my end perhaps?

**Comment by 13-05** [comment]
I'll close this as it's not inherently tied to `llama-rs`, rather my own system.

**Comment by philpax** [comment]
Yeah, I'm not sure. What CPU do you have?

**Comment by 13-05** [comment]
Pretty dated, `Intel Core i3-2310M` from 2011, dual core haha. But it builds rather quickly, so I don't think the issue would be tied to my CPU.

**Comment by philpax** [comment]
Ah yeah, that's likely your problem. I think it's using extensions that aren't compatible with your CPU - it doesn't seem like ggml handles that case during the build. You might be able to tinker around with ggml.c/h and our build.rs to get it working, but I'm not sure.

Even if you do get it working, I suspect the performance would be pretty poor - the actual inference process absolutely pushes CPUs to their limits.

**Comment by 13-05** [comment]
Hm, I guess that makes sense. I knew the performance would be poor if I got it up and running, I just wanted to play with it anyway haha. Thanks for your help though.

**Comment by 13-05** [comment]
Hey, @philpax! I don't know what changed, but I decided to try this again and it worked flawlessly! I just cloned the repo and did `cargo run --release --bin llama-cli -- repl --model-path ~/ggml-alpaca-7b-q4.bin`, and it's working wonderfully. It's obviously not all that fast, but it's not that slow either. Here's the result from just a minute or two (lol):
![image](https://user-images.githubusercontent.com/64506392/230627632-3e3a4c61-0b99-4d6b-88a5-6fced19b923c.png)
